### ISSUE TITLE
MethodPattern docs don't match code

### ISSUE THREAD

**Issue by mike-solomon** [openrewrite/rewrite-docs#160]
https://github.com/openrewrite/rewrite/blob/main/rewrite-java/src/main/java/org/openrewrite/java/MethodMatcher.java#L51

does not match

https://docs.openrewrite.org/reference/method-patterns#examples

Reported by: https://twitter.com/rotnroll666/status/1646809210695417857

**Comment by michael-simons** [comment]
Thanks a ton, Mike, much appreciated.

**Comment by timtebeek** [comment]
And thanks for reporting this issue @michael-simons ; anything in particular you were looking to use OpenRewrite for? Let us know if there's anything we can help with; we also have a Slack if you'd like to discuss more.

**Comment by michael-simons** [comment]
Yes, actually. Thanks @timtebeek 

I’m trying to figure out if it’s possible to
- find all classes annotated with an annotation 
- That have a field of type x and that field annotated with another pair of annotations 
- Change that type x to y
- Find all interfaces that have type parameters <changedType,x> and change that to <changedType,y>

in short: I wanna provide a recipe to change the id type of Spring Data entities and repositories

**Comment by timtebeek** [comment]
Intuitively I'd say that's possible; perhaps @kunli2 or @BoykoAlex can chime in with specifics, as they do this a little more often.
We can also move this discussion to a https://github.com/openrewrite/rewrite-spring issue or PR, if this is something you'd consider for inclusion with [our other Spring Data recipes](https://github.com/openrewrite/rewrite-spring/tree/main/src/main/java/org/openrewrite/java/spring/data).

**Comment by mike-solomon** [comment]
I posted about this in our company Slack and one of the devs had this to say:

I do notice that in our docs we say that `*` will match any one method argument, but in code `*` actually matches any character other than `..` Which means that a pattern like
`com.Foo foo(*)` would match a `String` but not `java.util.String`.

The current approach lets you do things like match arguments based on package, so something like `java.util.*`. But in all our recipes `MethodMatcher` we have not yet wanted to do that even once (that I know of).  But we've used `*` to mean "any one argument" many times, and all of those usages have a subtle bug where they would not match if the method is passed a fully-qualified name.

To the extent the code differs from the documentation it's probably our code that has unintuitive behavior rather than the docs needing correction. I believe the code should be brought in line with what we describe in the docs.

===

As it sounds like the docs are currently correct and code work needs to be done, I've opened this issue to track this instead: https://github.com/openrewrite/rewrite/issues/3138
